### ISSUE TITLE
Alignment with thin or thick characters

### ISSUE THREAD

**Issue by cpitclaudel** [Malabarba/smart-mode-line#169]
Pushing the function name marker of `which-function-mode` is great. This doesn't work too well with unicode in minor mode strings. For example with 

```
(diminish 'aggressive-indent-mode "↹") ;; Uses a serif fallback font
```

or 

```
(diminish 'nameless-mode " :") ;; Thin space
```

the `full`-width feature gets confused, because counting spaces is not enough.

Would this work using a pixel display spec and measuring strings in pixels?

**Comment by Malabarba** [comment]
> Would this work using a pixel display spec and measuring strings in pixels?

Yes, but I don't think there's any way to do that, unfortunately. Emacs has some pixel-width functions, but AFAIK they only apply to text written in buffer, you can't use these functions to canculate the pixel-width of strings. 

Of course, I'd be happy to be proven wrong. :-)

**Comment by cpitclaudel** [comment]
Would something like this do?

```
(save-window-excursion
  (with-temp-buffer
    (insert (format-mode-line mode-line-format))
    (set-window-buffer (selected-window) (current-buffer))
    (window-text-pixel-size nil (point-min) (point-max))))
```

**Comment by Malabarba** [comment]
My only fear is that it might be a little slow. The mode-line might be updated dozens of times per second if the user just holds down `C-n`, and if there's anything slow here the scrolling will be laggy.

It's certainly worth a try. Even if it's slow, we could use this mode only when there are variable-width characters on the string.

**Comment by cpitclaudel** [comment]
It'll be a little while before I have time to try to come up with a patch, so feel free to go ahead :) We could try it with spinner.el and see if it works with that.

**Comment by cpitclaudel** [comment]
How do you detect if there are variable-width characters?

**Comment by Malabarba** [comment]
> It'll be a little while before I have time to try to come up with a patch, so feel free to go ahead :)

I was going to say the same :)

> How do you detect if there are variable-width characters?

There might be a char-code property for that, but I'm not sure.

**Comment by cpitclaudel** [comment]
>> It'll be a little while before I have time to try to come up with a patch, so feel free to go ahead :)
>
> I was going to say the same :)

:)

>> How do you detect if there are variable-width characters?
>
> There might be a char-code property for that, but I'm not sure.

I think it's going to be font-dependent, so I'd be a bit surprised if it did.
But what we could do is cache widths: the minor modes list is the one that's going to have funky characters most of the time, so we could keep an LRU cache of mode-line -> widths. But that's getting really convoluted.

**Comment by Malabarba** [comment]
And so the complexity grows. :-)

TBH, sml could really use some refactorings. It was my very first package, and parts of its code are embarassing to say the least. :-)
